polynote-frontend/polynote/ui/component/comment.ts

```diff
@@ -0,0 +1,519 @@
+    const uuid = uuidv4();
+    return new CellComment(
+        uuid,
+        range,
+        author,
+        createdAt,
+        content,
+    )
+}
+
+// Comment Handler is in charge of comments for a particular cell
+export class CommentHandler extends UIMessageTarget {
+    private comments: Record<CommentID, CellComment> = {}; // holds all comments (root and replies)
+    private commentUIs: Record<string, CommentUI> = {};    // lookup table for comment UIs (keys are ranges, serialized as strings)
+    private commentRoots: Record<string, CommentID> = {};  // lookup table for range (serialized as string) to root comments
+
+    constructor(readonly cellId: number, readonly editor: IStandaloneCodeEditor) {
+        super();
+
+        this.subscribe(CreateComment, (cellId, comment) => {
+            this._add(comment);
+        });
+        this.subscribe(UpdateComment, (cellId, commentId, range, content) => {
+            this._update(commentId, range, content);
+        });
+        this.subscribe(DeleteComment, (cellId, commentId) => {
+            this._delete(commentId);
+        });
+
+    }
+
+    // placeholder for avatars
+    private static fetchAvatar(author: string) {
+        return undefined;
+    }
+
+    private _add(comment: CellComment) {
+        this.comments[comment.uuid] = comment;
+        const maybeRootId = this.commentRoots[comment.range.asString];
+        if (maybeRootId === undefined || this.comments[maybeRootId].createdAt > comment.createdAt) {
+            this.commentRoots[comment.range.asString] = comment.uuid;
+        }
+    }
+
+    add(comment: CellComment) {
+        this._add(comment);
+        if (!this.commentUIs[comment.range.asString]) {
+            this.initializeUI(comment.range);
+        }
+
+        this.commentUIs[comment.range.asString].add(comment);
+    }
+
+    private _update(commentId: CommentID, range: PosRange, content: string) {
+        const prev = this.comments[commentId];
+        const upd = new CellComment(prev.uuid, range, prev.author, prev.createdAt, content);
+
+        // we need to update some things if the range has been changed and this is a root comment.
+        if (range !== prev.range && this.commentRoots[prev.range.asString] === commentId) {
+            // of course, we need to reset the comment root
+            delete this.commentRoots[prev.range.asString];
+            this.commentRoots[range.asString] = commentId;
+
+            // next, update the comment UIs
+            const ui = this.commentUIs[prev.range.asString];
+            delete this.commentUIs[prev.range.asString];
+            this.commentUIs[range.asString] = ui;
+            ui.range = range;
+
+            // finally, update all the other comments.
+            Object.values(this.comments).filter(c => c !== prev).forEach(c => {
+                if (c.range.equals(prev.range)) {
+                    this.publish(new UpdateComment(this.cellId, c.uuid, range, c.content));
```

Comment:
hmm, I think this should happen directly on the backend. When there is a `Delete` or `Insert` we should update all the comments ranges against the edit.